src/lib.rs

```diff
@@ -596,33 +596,34 @@ mod impl_rustc {
 mod impl_serde {
     extern crate serde;
     use self::serde::{Serialize, Serializer, Deserialize, Deserializer};
-    use self::serde::de::Error;
+    use self::serde::de::{Error, Unexpected};
     use super::{OrderedFloat, NotNaN};
     use num_traits::Float;
+    use std::f64;
 
     impl<T: Float + Serialize> Serialize for OrderedFloat<T> {
-        fn serialize<S: Serializer>(&self, s: &mut S) -> Result<(), S::Error> {
+        fn serialize<S: Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
             self.0.serialize(s)
         }
     }
 
     impl<T: Float + Deserialize> Deserialize for OrderedFloat<T> {
-        fn deserialize<D: Deserializer>(d: &mut D) -> Result<Self, D::Error> {
+        fn deserialize<D: Deserializer>(d: D) -> Result<Self, D::Error> {
             T::deserialize(d).map(OrderedFloat)
         }
     }
 
     impl<T: Float + Serialize> Serialize for NotNaN<T> {
-        fn serialize<S: Serializer>(&self, s: &mut S) -> Result<(), S::Error> {
+        fn serialize<S: Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
             self.0.serialize(s)
         }
     }
 
     impl<T: Float + Deserialize> Deserialize for NotNaN<T> {
-        fn deserialize<D: Deserializer>(d: &mut D) -> Result<Self, D::Error> {
+        fn deserialize<D: Deserializer>(d: D) -> Result<Self, D::Error> {
             T::deserialize(d).and_then(|v| {
                 NotNaN::new(v)
-                    .map_err(|_| <D::Error as Error>::invalid_value("value cannot be NaN"))
+                    .map_err(|_| <D::Error as Error>::invalid_value(Unexpected::Float(f64::NAN), &"NaN"))
```

Comment:
> Mm, but testing with string comparison seems brittle and relies on serde's internal error string formatting.

Exactly the opposite. Testing with string comparison is the only way to test what the user is going to see. If you had written a test saying the error should be equal to `InvalidValue(Float(f64::NAN), "NaN")` it wouldn't have done us any good right?

> Does serde_test come with a Deserializer with an error enum that allows you to determine what kind of error it is and the parameters that were passed to it?

It comes with a Deserializer and an error enum that allows you to determine what error message would be displayed to the user.